### PR TITLE
Support kwargs in minimal_fx_tracer

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
The traced fn can now take keyword arguments, mirroring torch.export's pattern: (args, kwargs) is bundled as a single pytree and the kwarg key list is stored on TracedResult so run_traced reorders runtime kwargs to trace-time order (with a clear error on key mismatch).

Authored with Claude.

### CODE REVIEW

File: torchtitan/experiments/graph_trainer/make_fx_tracer.py [hunk 1]
```diff
@@ -503,7 +509,13 @@ def run_traced(
...
     """
     state_flat = list(state.values())
-    user_args_flat, _ = pytree.tree_flatten(list(args))
+    if set(kwargs) != set(traced_result.user_kwarg_keys):
+        raise ValueError(
+            f"kwarg key mismatch: got {list(kwargs)}, "
+            f"expected {list(traced_result.user_kwarg_keys)}"
+        )
+    kwargs = {k: kwargs[k] for k in traced_result.user_kwarg_keys}
+    user_args_flat, _ = pytree.tree_flatten((args, kwargs))
```
Comment (ydwu4) [inline torchtitan/experiments/graph_trainer/make_fx_tracer.py]: We could probably directly compare the stored in_spec vs in_spec at runtime so we could save the user_kwarg_keys in TracedResult.

File: torchtitan/experiments/graph_trainer/make_fx_tracer.py [hunk 2]
```diff
@@ -551,18 +569,23 @@ def run_traced_train_step(
...
+    of that name must be passed via the underlying ``fn``'s positional args.
+    """
 
     if not isinstance(module, nn.Module):
         raise ValueError(
             "run_traced_train_step requires args[0] to be an nn.Module, "
             f"got {type(module).__name__}."
         )
-    if any(isinstance(arg, nn.Module) for arg in args):
+    if any(isinstance(arg, nn.Module) for arg in (*args, *kwargs.values())):
```
Comment (ydwu4) [inline torchtitan/experiments/graph_trainer/make_fx_tracer.py]: Nit:the nn module check appears twice. seems can be extracted out as a helper.

File: torchtitan/experiments/graph_trainer/make_fx_tracer.py (hunk 1 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/make_fx_tracer.py]: +1, input_spec seems more general. 

also, I would consider runtime overhead. This is in the critical perf path.

File: torchtitan/experiments/graph_trainer/make_fx_tracer.py (hunk 2 above)
Comment (SherlockNoMad) [reply torchtitan/experiments/graph_trainer/make_fx_tracer.py]: +1

**Review by SherlockNoMad — APPROVED** [review]
you can land when you collect Yidi's stamp.

File: torchtitan/experiments/graph_trainer/make_fx_tracer.py [hunk 3]
```diff
@@ -501,9 +520,23 @@ def run_traced(
...
+
+    When ``_validate_runtime=True``, kwargs are validated and reordered against
+    the trace-time spec (raises on missing/extra keys). Off by default to keep
+    the per-step path overhead-free; the caller must pass kwargs in trace-time
+    order.
     """
     state_flat = list(state.values())
-    user_args_flat, _ = pytree.tree_flatten(list(args))
+    if _validate_runtime:
+        expected_kwarg_keys = traced_result.user_inputs_spec.child(1).context
```
Comment (ydwu4) [inline torchtitan/experiments/graph_trainer/make_fx_tracer.py]: ughh... i remember user_inputs_spec == user_args_spec will check the dictionary keys. We don't need to actually unroll the context i think?

File: torchtitan/experiments/graph_trainer/make_fx_tracer.py [hunk 4]
```diff
@@ -295,6 +307,7 @@ class TracedResult:
     num_flat_outputs: int
     output_subclass_layouts: dict[int, SubclassLayout]
     output_spec: pytree.TreeSpec
+    user_inputs_spec: pytree.TreeSpec
```
Comment (ydwu4) [inline torchtitan/experiments/graph_trainer/make_fx_tracer.py]: Can re-order the fields into the following in this PR?

```python
    gm: torch.fx.GraphModule

   # input related
    example_inputs: tuple[Any, ...]
    num_flat_inputs: int
    input_subclass_layouts: dict[int, SubclassLayout]
    user_inputs_spec: pytree.TreeSpec

    # output related
    num_flat_outputs: int
    output_subclass_layouts: dict[int, SubclassLayout]
    output_spec: pytree.TreeSpec


    # state related
    state_fqns: list[str]
```

Need a follow up PR: 

Can we do a full auditing of the fields in this dataclass because i'm not sure how these are related to each other and why they're useful from the documentation.

- example_inputs: should it actually be flat_example_inputs? Can they be accessed from ph.node.meta["val"]? 
- num_flat_inputs: how is this related to example_inputs?
- input_subclass_layout: is it on flat input?
- user_inputs_spec: pytree.TreeSpec

There seems to be a clear structure for both inputs and outputs: num, subclass_layouts, tree_spec. Might worth adding a level of abstraction.

File: torchtitan/experiments/graph_trainer/make_fx_tracer.py (hunk 3 above)
Comment (ydwu4) [reply torchtitan/experiments/graph_trainer/make_fx_tracer.py]: I see the problem. we could have valid call but with different pytree spec:

```python
fn(a, b, c)
fn(a, b, c=c)
```
will have different input_spec for fn but they're both valid.

To address this, i think we should do an args, kwargs binding/normalization first before pytree flatten.  In this way, if the binding fails that means the function is invoked in a way that's not supported.

File: torchtitan/experiments/graph_trainer/make_fx_tracer.py (hunk 4 above)
Comment (tugsbayasgalan) [reply torchtitan/experiments/graph_trainer/make_fx_tracer.py]: Yeah sounds good. For this PR, i will just reorder the fields.

File: torchtitan/experiments/graph_trainer/make_fx_tracer.py [hunk 5]
```diff
@@ -272,29 +272,50 @@ def extract_module_state(mod: nn.Module) -> dict[str, torch.Tensor]:
...
+
+    # output related
     num_flat_outputs: int
     output_subclass_layouts: dict[int, SubclassLayout]
     output_spec: pytree.TreeSpec
+
+    # state related
+    state_fqns: list[str]
+
     tensor_input_indices: list[int] = field(default_factory=list)
```
Comment (ydwu4) [inline torchtitan/experiments/graph_trainer/make_fx_tracer.py]: lol, what about this one! I didn't notice we have a tensor_input_indices.

File: torchtitan/experiments/graph_trainer/make_fx_tracer.py
Comment (ydwu4) [inline torchtitan/experiments/graph_trainer/make_fx_tracer.py:592]: Doesn't need to happen in this PR. 


Just noticed that this might cause kwarg collision e.g imagine kwargs consists of validate_module_fqns.

The correct way of handling this is just like minimal_fx_tracer, return a callable instead.
```python
run_traced(traced_result, module, validate_module_fqns)(*args, **kwargs)
```
, in that case, might as well as change run_traced to be replay():
```python
replay(traced_result, module, validate_module_fqns)(*args, **kwargs)
```

File: torchtitan/experiments/graph_trainer/tests/test_trace_module.py [hunk 6]
```diff
@@ -486,6 +486,92 @@ def test_core_explicit_state_train_step(self):
...
+            with torch.nn.utils.stateless._reparametrize_module(model, state):
+                return model(tokens) * scale
+
+        state = extract_module_state(model)
+        traced = minimal_fx_tracer(forward)(state, tokens, scale=scale)
+        out_ref = forward(state, tokens, scale=scale)
+        out_traced = run_traced(traced, state, tokens, scale=scale)
+        self.assertTrue(torch.equal(out_ref, out_traced))
+
+    def test_kwargs_runtime_reorder_raises(self):
```
Comment (ydwu4) [inline torchtitan/experiments/graph_trainer/tests/test_trace_module.py:504]: Add a TODO here so when upstream pytree dict key sorting issue is fixed, the author could be comfortable to removes this.

File: torchtitan/experiments/graph_trainer/make_fx_tracer.py
Comment (tugsbayasgalan) [reply torchtitan/experiments/graph_trainer/make_fx_tracer.py:592]: Yeah hahaha. I will do this in https://github.com/pytorch/torchtitan/pull/3164/changes#diff-8fe284997e4eaba06224a6516c1222b5bc1c4e68dac9489de2bf91de4306b574

File: torchtitan/experiments/graph_trainer/make_fx_tracer.py (hunk 5 above)
Comment (tugsbayasgalan) [reply torchtitan/experiments/graph_trainer/make_fx_tracer.py]: It is default. But yeah i moved it by not making it default. This seems precompile related.

File: torchtitan/experiments/graph_trainer/tests/test_trace_module.py (hunk 6 above)
Comment (tugsbayasgalan) [reply torchtitan/experiments/graph_trainer/tests/test_trace_module.py:504]: added in the make_fx_tracer.py

**Review by ydwu4 — APPROVED** [review]
